Support a basic devicetree toolchain

jacky8hyf opened the pull request:
This includes:

- An autodetected toolchain from host dtc and fdtoverlay
- Ability to customize the toolchain

Gansito144 commented on devicetree/private/autodetected_toolchain_repo.bzl:
```diff
@@ -0,0 +1,110 @@
+        if tool_path:
+            repository_ctx.symlink(tool_path, tool_name)
+            snippet += """\
+    {tool_name} = {repr_tool_name},
```
nit: `tool_name_repr`?

Gansito144 commented on devicetree/toolchain.bzl:
```diff
@@ -19,36 +12,36 @@ def _to_manifest_path(ctx, file):
+        if not target:
+            continue
+        transitive_tool_files.append(target.files)
+        transitive_tool_runfiles.append(target[DefaultInfo].default_runfiles)
```
does this have the right indentation or is it a Github UI issue?

Gansito144 replied:
I see, it was a side by side diff thing displayed weirdly. Sorry for the noise.

jacky8hyf commented on devicetree/toolchain.bzl:
```diff
@@ -19,36 +12,36 @@ def _to_manifest_path(ctx, file):
-        target_tool_path = target_tool_path,
-        tool_files = tool_files,
+        files = tool_files,
+        runfiles = ctx.runfiles(transitive_files = tool_files).merge_all(transitive_tool_runfiles),
```
done

jacky8hyf commented:
Apart from addressing your comments, I also fixed a bug on "Add autodetected_toolchain" that it doesn't respects --autodetect_toolchain flag, because of the `register_toolchain(:all)`. I fixed it by moving the toolchain implementations in a private package.

Gansito144 commented on devicetree/toolchain.bzl:
```diff
@@ -61,17 +58,18 @@ def _devicetree_toolchain_impl(ctx):
+            cfg = "target",
+            executable = True,
+        )
+        for name, doc in TOOLCHAIN_TOOLS.items()
```
Is this here to keep everything in sync? I think DAMP could apply here, because sometimes these rule definitions serve as a documentation (even when the autogenerated one displays the values correctly).

Gansito144 reviewed (approved):
I minor comment, but otherwise LGTM.
